[LoopVectorizer] Allow partial reductions to be made in predicated loops

JamesChesterman opened the pull request:
Does a select on the input rather than the output. This way the mask has the same number of lanes as the other operand in the select instruction.

JamesChesterman committed:
[LoopVectorizer] Allow partial reductions to be made in predicated
loops

Does a select on the input rather than the output. This way the
mask has the same number of lanes as the other operand in the
select instruction.

JamesChesterman committed:
Move condition that checks whether current reduction is a a partial
reduction.

JamesChesterman commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -9876,8 +9871,9 @@ void LoopVectorizationPlanner::adjustRecipesForReductions(
               : std::nullopt;
-      NewExitingVPV =
-          Builder.createSelect(Cond, OrigExitingVPV, PhiR, {}, "", FMFs);
+      if (!isa<VPPartialReductionRecipe>(OrigExitingVPV->getDefiningRecipe()))
```
Done, I've put it as an extra condition in the if statement above it. This is because everything within this if statement seems to replace `OrigExistingVPV` with `NewExistingVPV`, which is just `OrigExistingVPV` but with a `Select` instruction being performed on it.

SamTebbs33 commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -9703,7 +9698,8 @@ void LoopVectorizationPlanner::adjustRecipesForReductions(
     auto *OrigExitingVPV = PhiR->getBackedgeValue();
     auto *NewExitingVPV = PhiR->getBackedgeValue();
-    if (!PhiR->isInLoop() && CM.foldTailByMasking()) {
+    if (!PhiR->isInLoop() && CM.foldTailByMasking() &&
```
A comment here explaining why partial reductions have a special case would be good.

JamesChesterman replied:
Done

MacDue commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -2442,28 +2442,41 @@ class VPReductionPHIRecipe : public VPHeaderPHIRecipe,
   VPPartialReductionRecipe *clone() override {
-    return new VPPartialReductionRecipe(Opcode, getOperand(0), getOperand(1),
-                                        getUnderlyingInstr());
+    return getNumOperands() == 3
+               ? new VPPartialReductionRecipe(Opcode, getOperand(0),
+                                              getOperand(1), getOperand(2),
+                                              getUnderlyingInstr())
+               : new VPPartialReductionRecipe(Opcode, getOperand(0),
+                                              getOperand(1), nullptr,
+                                              getUnderlyingInstr());
```
```suggestion
    return  new VPPartialReductionRecipe(Opcode, getOperand(0),
                                              getOperand(1), getMask()
                                              getUnderlyingInstr());
```

JamesChesterman replied:
Done, thank you.

SamTebbs33 commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -2442,28 +2442,41 @@ class VPReductionPHIRecipe : public VPHeaderPHIRecipe,
   VPPartialReductionRecipe *clone() override {
-    return new VPPartialReductionRecipe(Opcode, getOperand(0), getOperand(1),
-                                        getUnderlyingInstr());
+    return getNumOperands() == 3
+               ? new VPPartialReductionRecipe(Opcode, getOperand(0),
+                                              getOperand(1), getOperand(2),
+                                              getUnderlyingInstr())
+               : new VPPartialReductionRecipe(Opcode, getOperand(0),
+                                              getOperand(1), nullptr,
+                                              getUnderlyingInstr());
```
I think it would be cleaner to set a `VPValue*` variable to null or the mask then pass that to the constructor, rather than choosing the constructor to call.

SamTebbs33 replied:
Just saw Ben's better suggestion.

JamesChesterman replied:
Done, did Ben's suggestion

SamTebbs33 commented on llvm/test/Transforms/LoopVectorize/AArch64/partial-reduce-dot-product-neon.ll:
```diff
@@ -1596,7 +1596,7 @@ define i32 @not_dotp_predicated_pragma(i64 %N, ptr %a, ptr %b) {
 ; CHECK-INTERLEAVE1-NEXT:    [[INDEX:%.*]] = phi i64 [ 0, [[VECTOR_PH]] ], [ [[INDEX_NEXT:%.*]], [[PRED_LOAD_CONTINUE62:%.*]] ]
 ; CHECK-INTERLEAVE1-NEXT:    [[VEC_IND:%.*]] = phi <16 x i64> [ <i64 0, i64 1, i64 2, i64 3, i64 4, i64 5, i64 6, i64 7, i64 8, i64 9, i64 10, i64 11, i64 12, i64 13, i64 14, i64 15>, [[VECTOR_PH]] ], [ [[VEC_IND_NEXT:%.*]], [[PRED_LOAD_CONTINUE62]] ]
-; CHECK-INTERLEAVE1-NEXT:    [[VEC_PHI:%.*]] = phi <16 x i32> [ zeroinitializer, [[VECTOR_PH]] ], [ [[TMP180:%.*]], [[PRED_LOAD_CONTINUE62]] ]
+; CHECK-INTERLEAVE1-NEXT:    [[VEC_PHI:%.*]] = phi <4 x i32> [ zeroinitializer, [[VECTOR_PH]] ], [ [[PARTIAL_REDUCE:%.*]], [[PRED_LOAD_CONTINUE62]] ]
```
We actually want these tests to be partially-reduced now so you could remove the `not` part of the name (same for the other file).

JamesChesterman replied:
Done.
I also added the `!7` to both `@dotp_predicated` tests, which made them almost the same as the `@dotp_predicated_pragma` tests, so the pragma tests have been removed.

SamTebbs33 reviewed (commented):
Do you know why the other predication test isn't getting a partial reduction?

SamTebbs33 commented:
> Do you know why the other predication test isn't getting a partial reduction?

It actually looks like the other predicated test is missing the loop metadata to make it predicated. It should still get partially-reduced if you add the `!loop !7` metadata to it.

JamesChesterman committed:
Address comments, such as removing and renaming tests

SamTebbs33 commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -9698,6 +9698,11 @@ void LoopVectorizationPlanner::adjustRecipesForReductions(
     // beginning of the dedicated latch block.
     auto *OrigExitingVPV = PhiR->getBackedgeValue();
     auto *NewExitingVPV = PhiR->getBackedgeValue();
+    // Don't add selects here for partial reductions because the phi and partial
+    // reduction values have less vector elements than Cond. But, each operand
+    // in a select instruction needs to have the same number of vector elements,
+    // so the compiler would crash. Instead, a select, with the active lane
+    // mask, is applied to the inputs to the partial reduction.
```
This could probably be more concise, as something like "The output of a partial reductions has fewer lanes than the VF so don't emit a select, as its operands would have differing numbers of lanes. A partial reduction masks its input instead."

JamesChesterman committed:
Address comments. Add RUN line to test files. Change a comment.
Remove Mask as operand in Partial Reduction Recipe.
Instead just mask the input when creating the recipe.

JamesChesterman commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -2442,28 +2442,41 @@ class VPReductionPHIRecipe : public VPHeaderPHIRecipe,
   VPPartialReductionRecipe *clone() override {
-    return new VPPartialReductionRecipe(Opcode, getOperand(0), getOperand(1),
-                                        getUnderlyingInstr());
+    return getNumOperands() == 3
```
Done, added a RUN line with `-enable-epilogue-vectorization=true` to both test files.

MacDue replied:
Do we still need the epilogue test cases now that the changes to the clone code have been removed?

MacDue commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -8904,6 +8897,12 @@ VPRecipeBuilder::tryToCreatePartialReduction(Instruction *Reduction,
       isa<VPPartialReductionRecipe>(BinOpRecipe))
     std::swap(BinOp, Accumulator);
 
+  VPValue *Mask = getBlockInMask(Reduction->getParent());
+  if (Mask) {
+    VPValue *Zero =
+        Plan.getOrAddLiveIn(ConstantInt::get(Reduction->getType(), 0));
+    BinOp = Builder.createSelect(Mask, BinOp, Zero, Reduction->getDebugLoc());
+  }
```
Not sure, but should this be (or similar):
```suggestion
  if (CM.isPredicatedInst(Reduction)) {
    VPValue *Mask = getBlockInMask(Reduction->getParent());
    VPValue *Zero =
        Plan.getOrAddLiveIn(ConstantInt::get(Reduction->getType(), 0));
    BinOp = Builder.createSelect(Mask, BinOp, Zero, Reduction->getDebugLoc());
  }
```
Just because I see a `Trying to access mask for block without one.` assert in `getBlockInMask()`, and I wonder if it's possible to hit that somehow.

JamesChesterman replied:
Done.
The assert wasn't getting hit for the non-predicated tests, as they would return a mask of 0 rather than no mask at all. But I've added the extra condition for more safety. It needed to be a slightly different function inside the condition as this one was not working as required.

MacDue commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -2448,7 +2448,8 @@ class VPReductionPHIRecipe : public VPHeaderPHIRecipe,
 /// vector operand are added together and passed to the next iteration as the
 /// next accumulator. After the loop body, the accumulator is reduced to a
-/// scalar value.
+/// scalar value. If the mask operand is not nullptr then it is applied to the
+/// vector operand on each iteration.
```
Outdated comment? There's no mask operand anymore.

JamesChesterman replied:
Done

JamesChesterman committed:
Address comments on PR

Involves adding an extra if statement and changing a comment.

MacDue reviewed (approved):
I think the approach generally LGTM, though I'm inclined to say the extra epilogue vectorization tests should be removed, as they massively increase the diff (making it much harder to see the actual change). I think the original motivation for adding them has been removed from the PR (the `clone()` function). cc @fhahn 

If we do need some more tests I think it's best to reduce them down to checking only the relevant parts, to make them smaller and easier to understand. E.g. In this case, what you wanted to check is that the partial reduction in the `epilog.vector.body` has its inputs masked like the one in `vector.body` (with that we know the `clone()` function worked).

fhahn commented on llvm/test/Transforms/LoopVectorize/AArch64/partial-reduce-dot-product-neon.ll:
```diff
@@ -2,6 +2,7 @@
 ; RUN: opt -passes=loop-vectorize -enable-epilogue-vectorization=false -mattr=+neon,+dotprod -force-vector-interleave=1 -S < %s | FileCheck %s --check-prefixes=CHECK-INTERLEAVE1
 ; RUN: opt -passes=loop-vectorize -enable-epilogue-vectorization=false -mattr=+neon,+dotprod -S < %s | FileCheck %s --check-prefixes=CHECK-INTERLEAVED
 ; RUN: opt -passes=loop-vectorize -enable-epilogue-vectorization=false -mattr=+neon,+dotprod -force-vector-interleave=1 -vectorizer-maximize-bandwidth -S < %s | FileCheck %s --check-prefixes=CHECK-MAXBW
+; RUN: opt -passes=loop-vectorize -enable-epilogue-vectorization=true -mattr=+neon,+dotprod -S < %s | FileCheck %s --check-prefixes=CHECK-EPILOGUE
```
There is already a separate test file to check epilogue codegen for partial reductions: `llvm/test/Transforms/LoopVectorize/AArch64/partial-reduce-dot-product-epilogue.ll`.

Can you add the predicated case there instead?

JamesChesterman replied:
Done

JamesChesterman committed:
Address comments. Involves moving predicated test with epilogue
vectorisation to a separate test file.

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -8904,6 +8897,12 @@ VPRecipeBuilder::tryToCreatePartialReduction(Instruction *Reduction,
+  if (CM.blockNeedsPredicationForAnyReason(Reduction->getParent())) {
+    VPValue *Mask = getBlockInMask(Reduction->getParent());
+    VPValue *Zero =
+        Plan.getOrAddLiveIn(ConstantInt::get(Reduction->getType(), 0));
```
This assumes Add/Sub as opcodes to use 0 as neutral element. Can you add an assert to clarify this?

JamesChesterman replied:
Done, thank you for spotting that.

JamesChesterman committed:
Add assert statement to ensure the PR is an Add or Sub

This is because the neutral element in the mask is zero.

fhahn reviewed (approved):
LGTM, thanks
